### PR TITLE
Multiplayer commands fix

### PR DESCRIPTION
**Pull request by Limeth**
Call now and get 2 more bugfixes for free.

### CODE REVIEW

File: engine/src/main/java/org/terasology/logic/health/HealthSystem.java [hunk 1]
```diff
@@ -211,15 +211,9 @@ public String damage(@Sender EntityRef client, @CommandParam("amount") int amoun
...
-        ClientComponent clientComp = clientEntity.getComponent(ClientComponent.class);
-        clientComp.character.send(new DoHealEvent(100000, clientComp.character));
-        return "Health restored";
-    }
-
-    @Command(shortDescription = "Restores your health by an amount", runOnServer = true)
-    public void health(@Sender EntityRef client, @CommandParam("amount") int amount) {
+    @Command(shortDescription = "Restores your health by a specifiad amount (or fully)", runOnServer = true)
+    public void health(@Sender EntityRef client, @CommandParam(value = "amount", required = false) Integer amountParam) {
+        int amount = amountParam != null ? amountParam : 100000;
```
Comment (Cervator) [inline engine/src/main/java/org/terasology/logic/health/HealthSystem.java]: 100000 is a magic value some conditions might just exceed one day - suggest Integer.MAX_VALUE instead :-)

**Comment by Cervator** [comment]
Your ideas are intriguing to me and I wish to subscribe to your newsletter!

Looks great - was able to get "op" with headless, have two clients chatting with good behavior either from the main console or the chat console, get notification messages, use commands, and so on

Looks like this fixes #1482 :-)

I added one tiny comment about a magic number, but the changes look overall minor to me. I don't have much background to think about the pros and cons of using a String vs an Enum in one place.

@immortius any observations? Since this has been such an active topic I wanted to still run it by you, but am happy to go back to merging stuff like this that appears minor if you're happy :-)

File: engine/src/main/java/org/terasology/logic/notifications/NotificationMessageEvent.java [hunk 2]
```diff
@@ -54,17 +53,23 @@ public static NotificationMessageEvent newLeaveEvent(EntityRef client) {
...
-    public EntityRef getFrom() {
-        return from;
+    @Override
+    public String getFormattedMessage() {
+        return message;
     }
-    
+
     @Override
-    public Message getFormattedMessage() {
```
Comment (immortius) [inline engine/src/main/java/org/terasology/logic/notifications/NotificationMessageEvent.java]: Why change this?

**Comment by immortius** [comment]
I don't understand why this replaces the message type enums with string constants. That didn't need to be changed.

File: engine/src/main/java/org/terasology/logic/health/HealthSystem.java [hunk 3]
```diff
@@ -211,15 +211,9 @@ public String damage(@Sender EntityRef client, @CommandParam("amount") int amoun
...
-    @Command(shortDescription = "Restores your health to max", runOnServer = true)
-    public String health(@Sender EntityRef clientEntity) {
-        ClientComponent clientComp = clientEntity.getComponent(ClientComponent.class);
-        clientComp.character.send(new DoHealEvent(100000, clientComp.character));
-        return "Health restored";
-    }
-
-    @Command(shortDescription = "Restores your health by an amount", runOnServer = true)
-    public void health(@Sender EntityRef client, @CommandParam("amount") int amount) {
+    @Command(shortDescription = "Restores your health by a specifiad amount (or fully)", runOnServer = true)
```
Comment (immortius) [inline engine/src/main/java/org/terasology/logic/health/HealthSystem.java]: Specified is mispelt.

File: engine/src/main/java/org/terasology/logic/health/HealthSystem.java (hunk 1 above)
Comment (immortius) [inline engine/src/main/java/org/terasology/logic/health/HealthSystem.java]: I would suggest actually looking up the characters max health.

**Comment by immortius** [comment]
From what I can tell this fix breaks the message coloring for different message types.

**Comment by Limeth** [comment]
Yes, why have the additional complexity of specifying a message type, if it only provides a color? The color can be set by the FontColor.getColored bethod anyway.

**Comment by immortius** [comment]
It isn't just about providing color, it is also providing a type for filtering/display purposes. A view may be set up to only show chat messages for instance. And I prefer something a little less flimsy than strings be used.

Notably, you did not mention these changes in the PR description, and it isn't part of the bug fix.

**Comment by Cervator** [comment]
Went back to look at colors, not sure exactly which - player names are colored, auto-completion is colored, unknown command stuff is in red. I'm probably missing some :-)

Filters sound sensible

I did spot one surviving bug: When attached to a headless server without "op" right the client still doesn't get a permission denied error (for instance for "giveBlock dirt") - it just prints on the server's console, client side gets no output

**Comment by immortius** [comment]
Anyway, if possible can we split the changes to message type into a separate PR and I'll write up a full set of requirements for us to consider them in context of. Cheers.

**Comment by Cervator** [comment]
IIRC this came up on IRC the other day and the MessageType was involved in the command fix. Maybe something about serialization or missing type handlers, or what not, I could easily just not be remembering it right. In either case simplifying it (via String) was a possible fix to the multiplayer command bug, so tearing the two apart to handle separately might not be as easy as we'd like

With a little luck I'm just remembering wrong :-)

**Comment by immortius** [comment]
Well, it worked before so I'm sure commands could be changed to work again without changing MessageType.

It probably isn't a perfect system anyway though. The requirements I had in mind were:
* Messages have a type associated with them.
* New message types should be introducable by modules.
* Message types should be usable for filtering for different displays (e.g. console vs chat)
* Each message type should have a default color, that is the color of the message if not overridden by color tokens (or if their use is disallowed for that message type, which might be sensible for chat messages). This shouldn't be handled by individual message submitters - their should be a color per type and linked directly to the type. Individual displays may override this. May also be nice to allow the color to be overridden for the entire type by modules.

Using enums over strings has a couple of other advantages, including clarity (```void send(String, MessageType)``` vs ```void send(String, String)```), reduced possibility of error (cannot misspell an enum) and that the enums could have ids assigned to them up front and these ids can be transmitted rather than their string values (reduces bandwidth requirements for messages). The last cannot be readily done for string constants because they cannot be discovered up front.

**Comment by Limeth** [comment]
The problem was, that MessageType isn't an enum, but an interface. Then there was an enum implementing the interface. How would that be sent to the client? Also, how are Events sent to clients and vice versa?

**Comment by immortius** [comment]
Currently (or at least, before the command system rework), we don't ever send message type anywhere. There is an event class for each message type, and that provides the message - they type is created by the message event, never transmitted. So the question of how to serialize them is not relevant. They could be serialized by finding all enums implementing the interface via Reflections if need be.

Events are sent by being serialized using our persistence framework (same thing that is used to load and save the game, and to transmit entities) and then sent in serialized form. Essentially the sending of events is triggered by the annotations on the event (```@ServerEvent``` for events that should be actioned by the server (will only work if sent to a network entity), ```@OwnerEvent``` for events that should be sent to the player owning the entity the event is sent to, and ```@BroadcastEvent``` for events sent to all players).

**Comment by immortius** [comment]
Having a bit of a look into what is actually going on:
* The permissions are stored on the player's character, so they lose all their permissions when they die. Not related to the current issues, but I thought worth pointing out.
* All commands default to requiring "op" permission. I really think we should switch to default no permission. Chat commands currently require "op" permission because they don't specifically say the required permission is "".
* During command processing, messages should not be added directly to the console. The message events (primarily ConsoleMessage) should be sent to the calling client entity, which will result in them being sent across the network if necessary. This is missing in the permission handling.

**Comment by Cervator** [comment]
Bump - with #1496 we have the tiny fix option in PR form, but it'll conflict with this PR. @Limeth how would you like to proceed? Would you prefer tweaking this PR so it can be merged instead (if doable with minor easy work), or close it in favor of #1496 then make a new one later with any additional fixes from here?

**Comment by Limeth** [comment]
Replaced by #1496

**Pull request closed without merging**